routes/dao_coin_exchange_with_fees.go

```diff
@@ -0,0 +1,1549 @@
+		}
+		feeBaseUnitsByPkid[pkid] = feeBaseUnits
+		totalFeeBaseUnits, err = lib.SafeUint256().Add(totalFeeBaseUnits, feeBaseUnits)
+		if err != nil {
+			return nil, fmt.Errorf("HandleMarketOrder: Problem calculating total fee: %v", err)
+		}
+	}
+
+	// Validate that the totalFeeBaseUnits is less than or equal to the quote currency total
+	if totalFeeBaseUnits.Cmp(quoteCurrencyExecutedBeforeFeesBaseUnits) > 0 {
+		return nil, fmt.Errorf("HandleMarketOrder: Total fees exceed total quote currency")
+	}
+
+	// Precompute the total fee to return it later
+	marketTakerFeeBaseUnits := uint64(0)
+	for _, feeBaseUnits := range feeMap {
+		marketTakerFeeBaseUnits += feeBaseUnits
+	}
+	marketTakerFeeBaseUnitsStr := fmt.Sprintf("%d", marketTakerFeeBaseUnits)
+
+	// Now we have two possibilities...
+	//
+	// 1. buy
+	// The user is buying the base currency with the quote currency. In this case we can
+	// simply deduct the quote currency from the user's balance prior to executing the
+	// order, and then execute the order with remainingQuoteCurrencyBaseUnits.
+	//
+	// 2. sell
+	// In this case the user is selling the base currency for quote currency. In this case,
+	// we need to execute the order first and then deduct the quote currency fee from the
+	// user's balance after the order has been executed.
+	//
+	// Get a universal view to validate as we go
+	utxoView, err := fes.backendServer.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		return nil, fmt.Errorf("HandleMarketOrder: Error fetching mempool view: %v", err)
+	}
+	transactorPubkeyBytes, _, err := lib.Base58CheckDecode(req.TransactorPublicKeyBase58Check)
+	if err != nil || len(transactorPubkeyBytes) != btcec.PubKeyBytesLenCompressed {
+		return nil, fmt.Errorf("HandleMarketOrder: Problem decoding public key %s: %v",
+			req.TransactorPublicKeyBase58Check, err)
+	}
+	quoteCurrencyPkidBytes, _, err := lib.Base58CheckDecode(req.QuoteCurrencyPublicKeyBase58Check)
+	if err != nil || len(quoteCurrencyPkidBytes) != btcec.PubKeyBytesLenCompressed {
+		return nil, fmt.Errorf("HandleMarketOrder: Problem decoding public key %s: %v",
+			req.QuoteCurrencyPublicKeyBase58Check, err)
+	}
+	if isBuyOrder {
+		// For each trading fee we need to pay, construct a transfer txn that sends the amount
+		// from the transactor directly to the person receiving the fee.
+		transferTxns := []*lib.MsgDeSoTxn{}
+		for pkid, feeBaseUnits := range feeBaseUnitsByPkid {
+			receiverPkidBytes, _, err := lib.Base58CheckDecode(pkid)
+			if err != nil || len(receiverPkidBytes) != btcec.PubKeyBytesLenCompressed {
+				return nil, fmt.Errorf("HandleMarketOrder: Problem decoding public key %s: %v",
+					pkid, err)
+			}
+			// Try and create the TransferDaoCoin transaction for the user.
+			//
+			// TODO: Add ExtraData to the transaction to make it easier to report it as an
+			// earning to the user who's receiving the fee.
+			txn, err := fes.SendCoins(
```

Comment:
I don't think it's worth the added complexity. The reason it's hard is because you may have multiple recipients who need to get fees. If that's the case, then yes you can do it with a single transfer with DESO but you *can't* for DAO coin, which means you have to special-case stuff.

Reply:
alright that's fine - it just becomes more expensive to assess fees.